### ISSUE TITLE
Key Code List

### ISSUE THREAD

**Issue by TomFaulkner** [david-janssen/kmonad#24]
I looked around, but I couldn't find one. Is there a list of key codes?

in the documentation and examples I didn't see the name for `Print Screen`, `Pause Break`, volume up and down, and mute.

**Comment by TomFaulkner** [comment]
Nevermind this, I found the link to the header file.

Thanks for the hard work in this, I may actually use my laptop keyboard at some point now. I haven't been able to since I went Colemak.

**Comment by david-janssen** [comment]
Glad to see you found it. And thanks for the feedback!

**Comment by TomFaulkner** [comment]
While I have your attention though, is there a way to do auto shift like QMK has? (holding any letter or number that isn't already assigned to tap for letter and hold for something else gives and that key, That is I hold `a` and get a capital A, but I can hold `c` and it works as Super.

**Comment by david-janssen** [comment]
Hey Tom,

I am not familiar with AutoShift, so maybe I am interpreting it wrong, but I think you could duplicate the intended behavior using a TapHold key. It takes 3 arguments: a delay, a button to tap if released within the delay, and a button to press the moment that delay is exceeded (and then release when the release is detected).  You can find more documentation here: https://github.com/david-janssen/kmonad/blob/master/doc/syntax_guide.md#tap-hold

So you could define:
```
@foo = TH 300 a A
@bar = TH 400 c lalt 
```

(There is no 'Super' key in the linux keycode headers)

**Comment by TomFaulkner** [comment]
Yeah, that is pretty much it. Would just have to do that with every key.

As for `super` `lmeta` seems to be the equivalent.

**Comment by david-janssen** [comment]
Out of curiosity: how does this work with QMK. You enable auto-shift for the entire driver or just for a layer?

**Comment by TomFaulkner** [comment]
From ZSA`s Oryx configuration tool:

Get rid of your Shift key!

> Tap a key and you get its character. Tap a key, but hold it slightly longer and you get its shifted state. Voilà! No shift key needed! Close

> Why Auto Shift?

> Many people suffer from various forms of RSI. A common cause is stretching your fingers again and again over long distances. The pinky does that all too often when reaching for the Shift key. Auto Shift is one way to alleviate that problem.

How Does It Work?

> When you tap a key briefly, it sends out its usual keystroke. When you tap it for a longer period of time, it sends out its shifted state. Adjusting that period of time for your personal typing style is important. Watch the video to see the exact process, which also involves assigning some special keys, not just the settings on this screen.
Are There Limitations to Auto Shift?

> Yes, unfortunately.

> Key repeat will cease to work. For example, before if you wanted 20 'a' characters, you could press and hold the 'a' key for a second or two. This no longer works with Auto Shift because it is timing your depressed time instead of emitting a depressed key state to your operating system.
> You will have characters that are shifted when you did not intend on shifting, and other characters you wanted shifted, but were not. This simply comes down to practice. As we get in a hurry, we think we have hit the key long enough for a shifted version, but we did not. On the other hand, we may think we are tapping the keys, but really we have held it for a little longer than anticipated.

https://configure.ergodox-ez.com/planck-ez/layouts/jZ0yd/latest/config

**Comment by david-janssen** [comment]
Hmmmm, interesting. At the moment most of this behavior can be simply duplicated by creating tap-hold buttons. That is very laborious though, so it's not optimal. I've been thinking about how best to structure the configuration language, and if you could just map functions over your keymaps, you could remove nearly all the boiler-plate. No internal implementation stuff needs to change.

You've given me food for thought, check back in a few months :-)

**Comment by TomFaulkner** [comment]
Will do. Thanks!

**Comment by Herschenglime** [comment]
Hello! I really wanted autoshift so I ended up taking the time to implement it in a kmonad layout as you described in the upper post and it works great! I haven't added anything else by way of layers, so I'd love to be able to share it for other people to use to implement this functionality. I'm not 100% sure how I'd go about doing that, perhaps through the user config section?

I am a bit confused on how to use the output function in defcfg though. If I leave the optional argument empty, kmonad fills the terminal output with blank lines until I kill it. I don't use a compose key so I have no need for the config shown in the tutorial.kbd file.

Additionally, I'm running into a problem where I lose trackpad input while kmonad is running. Once I stop the command, it takes around 30 seconds for it to come back. if i plug in an external mouse or use the touchscreen, it's fine. I'm using an integrated laptop keyboard, so is it possible that the devices are tied together in some way and kmonad's control over keyboard input is affecting the touchpad as well?

Thank you so much for your help! I really appreciate what you all are doing with this project.

**Comment by Herschenglime** [comment]
Turns out the culprit for the loss of touchpad input had to do with a KDE setting to disable touchpad while typing. I guess writing out that whole comment sparked the correct idea in my brain or something, heh. 

Also in the process of trying to edit my first comment, I discovered that you can attach files. I will do that here. Thanks again!

**Comment by david-janssen** [comment]
First of all: happy to hear you're having a good time with KMonad. If you want to include an illustration of how to implement auto-shift using KMonad, I'd be happy to merge a pull-request for your own keymap into the user keymaps. That directory will probably be splitting off into its own repo under the kmonad organization sometime soonish.

> It seems that running kmonad also somehow interrupts fusuma which I use for touchpad gestures.

That would surprise me. The only thing we do on Linux to grab a keyboard is to do an Linux `ioctl` grab on a particular device under `/dev/input`. Unless fusuma needs that source of keyboard events, they should not be interfering. Are you sure it's not something else?

> If I leave the optional argument empty, kmonad fills the terminal output with blank lines until I kill it.

Yah, what happens is that you press enter on your input keyboard to start KMonad. We then grab that keyboard so quickly that the release is never registered. All our outputs are output via a different, simulated keyboard, so the input keyboard stays in the 'enter-is-pressed' mode for the OS. Either launching kmonad like this: `sleep 0.5 && kmonad ...` or else using the `-w` flag on newer builds to specify a delay in startup so that you have time to release all the keys on the input keyboard should fix this issue.

This should be better documented, and we're moving towards it.

**Comment by Herschenglime** [comment]
Thanks for the quick reply! I'm still not super familar with github so I'll have to look into how pull requests work before I can get that shared.

Thanks for the sleep recommendation, I think I'll make an alias with the sleep period baked in to avoid the issue. It seems the fusuma interruption had to do with enter being stuck in the pressed down state. When this isn't occurring, it works fine. 

What is the id of the simulated keyboard that kmonad uses? It seems the aforementioned KDE option to disable touchpad while typing only watches the built-in keyboard. When the enter key isn't stuck in a pressed-down state, I have full use of my touchpad, but it also doesn't disable while typing with kmonad active. Perhaps somewhere in the KDE config files, I can find where the keyboard that gets disabled is set and change it to the virtual keyboard so I can continue to make use of this feature.


Thanks again for your help!

**Comment by david-janssen** [comment]
> What is the id of the simulated keyboard that kmonad uses? 

We use the `uinput` subsystem for linux, so that would be `/dev/uinput`.

> but it also doesn't disable while typing with kmonad active.

That's an interesting interaction I hadn't considered yet: simulated keyboard fails to trigger touchpad disabling in X11 :-).

We could even include touchpad disabling in KMonad... the moment you press a key it disables the touchpad, and won't reenable it until you've stopped typing for 'n' milliesconds... Not conceptually too difficult. Very much a `someday` issue, but interesting.

**Comment by Herschenglime** [comment]
I ended up speaking with the libinput people to get it figured out, which can be seen [here](https://gitlab.freedesktop.org/libinput/libinput/-/issues/617). In short though, to get disable while typing working, I had to create a file at `/etc/libinput/local-overrides.quirks` with
```
[My KMonad Output]
MatchUdevType=keyboard
MatchBus=usb
MatchName=My KMonad output
AttrKeyboardIntegration=internal
```
as its contents and reboot. The name of the quirk in the brackets can be changed, and the contents of MatchName is whatever you chose to name your uinput sink in your kmonad config file. This fix works by telling libinput that the kmonad output is an internal keyboard, making it also contribute to disable while typing. Since this doesn't actually interact with official libinput API's, it's not guaranteed to keep working, but for now it does exactly what I want.

Thanks for your help!
